controllers/mr_inferenceservice_controller.go

```diff
@@ -0,0 +1,414 @@
+		rp := existingISVC.DeepCopy()
+
+		_, okIsId := desiredISVC.Labels[constants.ModelRegistryInferenceServiceIdLabel]
+		_, okRegisteredModelId := desiredISVC.Labels[constants.ModelRegistryRegisteredModelIdLabel]
+		_, okModelVersionId := desiredISVC.Labels[constants.ModelRegistryModelVersionIdLabel]
+
+		if okIsId {
+			rp.Labels[constants.ModelRegistryInferenceServiceIdLabel] = desiredISVC.Labels[constants.ModelRegistryInferenceServiceIdLabel]
+		} else {
+			delete(rp.Labels, constants.ModelRegistryInferenceServiceIdLabel)
+		}
+
+		if okRegisteredModelId {
+			rp.Labels[constants.ModelRegistryRegisteredModelIdLabel] = desiredISVC.Labels[constants.ModelRegistryRegisteredModelIdLabel]
+		} else {
+			delete(rp.Labels, constants.ModelRegistryRegisteredModelIdLabel)
+		}
+
+		if okModelVersionId {
+			rp.Labels[constants.ModelRegistryModelVersionIdLabel] = desiredISVC.Labels[constants.ModelRegistryModelVersionIdLabel]
+		} else {
+			delete(rp.Labels, constants.ModelRegistryModelVersionIdLabel)
+		}
+
+		if err = r.client.Update(ctx, rp); err != nil {
+			return
+		}
+	}
+
+	if delta.IsRemoved() {
+		log.V(1).Info("Delta found", "delete", existingISVC.GetName())
+		if err = r.client.Delete(ctx, existingISVC); err != nil {
+			return
+		}
+	}
+	return nil
+}
+
+func (r *ModelRegistryInferenceServiceReconciler) getOrCreateServingEnvironment(log logr.Logger, mr api.ModelRegistryApi, namespace string) (*openapi.ServingEnvironment, error) {
+	servingEnvironment, err := mr.GetServingEnvironmentByParams(&namespace, nil)
+	if err != nil {
+		log.Info("ServingEnvironment not found, creating it..")
+		servingEnvironment, err = mr.UpsertServingEnvironment(&openapi.ServingEnvironment{
+			Name: &namespace,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("unable to create ServingEnvironment: %w", err)
+		}
+	}
+	return servingEnvironment, nil
+}
+
+// createMRInferenceService create a new model registry InferenceService resource based on provided input
+func (r *ModelRegistryInferenceServiceReconciler) createMRInferenceService(
+	log logr.Logger,
+	mr api.ModelRegistryApi,
+	isvc *kservev1beta1.InferenceService,
+	servingEnvironmentId string,
+	registeredModelId string,
+	modelVersionId string,
+) (*openapi.InferenceService, error) {
+	modelVersionIdPtr := &modelVersionId
+	if modelVersionId == "" {
+		modelVersionIdPtr = nil
+	}
+
+	isName := fmt.Sprintf("%s/%s", isvc.Name, isvc.UID)
+
+	log.Info("Creating new model registry InferenceService", "name", isName, "registeredModelId", registeredModelId, "modelVersionId", modelVersionId)
+	is := &openapi.InferenceService{
+		Name:                 &isName,
+		ServingEnvironmentId: servingEnvironmentId,
+		RegisteredModelId:    registeredModelId,
+		ModelVersionId:       modelVersionIdPtr,
+		Runtime:              isvc.Spec.Predictor.Model.Runtime,
+		DesiredState:         openapi.INFERENCESERVICESTATE_DEPLOYED.Ptr(),
+	}
+
+	return mr.UpsertInferenceService(is)
```

Comment:
That's great catch, I wasn't considering that case - as is it would have failed trying to create an InfService in Model Registry as already existing another one with the same name.

Fixing this logic by checking the existance of the record before creating it.

Reply:
Fixed.